### ISSUE TITLE
AsyncLocalStorage doesn't work with async/await

### ISSUE THREAD

**Issue by Janpot** [stackblitz/webcontainer-core#1169]
### **Description of Bug**

`AsyncLocalStorage` doesn't work well with `async`/`await`

### **Steps to Reproduce**

1. Go to https://stackblitz.com/edit/stackblitz-starters-pkxqds?file=index.mjs
    ```tsx
    import { AsyncLocalStorage } from 'node:async_hooks';

    const asyncLocalStorage = new AsyncLocalStorage();

    asyncLocalStorage.run('hello', () => {
      console.log('store before:', asyncLocalStorage.getStore());
      setTimeout(() => {
        console.log('store after:', asyncLocalStorage.getStore());
      }, 1);
    });

    asyncLocalStorage.run('hello', async () => {
      console.log('store async before:', asyncLocalStorage.getStore());
      await new Promise((r) => setTimeout(r, 1));
      console.log('store async after:', asyncLocalStorage.getStore());
    });
    ```
3. Run `node index.mjs`
4. Observe output 
    ```
    store before: hello
    store async before: hello
    store after: hello
    store async after: undefined
    ````
5. Running the same script locally produces output 
    ```
    store before: hello
    store async before: hello
    store after: hello
    store async after: hello
    ````
   (Running under Node.js v16.20.0 on MacOS)

### **Expected Behavior**

[`AsyncLocalStorage`](https://nodejs.org/api/async_context.html#class-asynclocalstorage) in stackblitz works with `async`/`await`.

**Comment by Nemikolh** [comment]
Hi @Janpot ! Thanks for opening this issue!

You are correct! `AsyncLocalStorage` does not currently works in StackBlitz. This is because we don't really have an implementation for async hooks and promise hooks. We have ideas about how to do that but we want to make sure it wouldn't impact performance too much.

We'll keep this issue open and keep you posted!

**Comment by Janpot** [comment]
Thank you @Nemikolh. Are there any recommended ways to detect whether my package is running under Stackblitz?

**Comment by Nemikolh** [comment]
Yes! You can use [`@webcontainer/env`](https://www.npmjs.com/package/@webcontainer/env). It exposes a function `isWebContainer()` that works both on Node and in WebContainer.

**Comment by pi0** [comment]
Related to https://github.com/nuxt/nuxt/issues/23032, it would be nice that if `AsyncLocalStorage` mock shows a one time warning while we are waiting for AsyncContext native support in browsers so that framework users implicitly relying on this API, consider that it is an unsupported platform feature instead of silently getting and error.

**Comment by birkskyum** [comment]
Is there any update on this?

**Comment by birkskyum** [comment]
Is there any update on this?

**Comment by pi0** [comment]
@birkskyum Last time I checked with the team, it needed native browser support. The best thing we might hope is to wait for [AsyncContext](https://github.com/tc39/proposal-async-context/) proposal to move forward and browsers to implement it.

**Comment by birkskyum** [comment]
In the proposal it states:

> Furthermore, the async/await syntax bypasses the userland Promises and makes it impossible for existing tools like [Zone.js](https://github.com/tc39/proposal-async-context/?tab=readme-ov-file#zonesjs) that [instruments](https://github.com/angular/angular/blob/main/packages/zone.js/STANDARD-APIS.md) the Promise to work with it without transpilation.

@mgechev do you have any on the horizon for Angular to sunset zone.js?

**Comment by tannerlinsley** [comment]
Has anyone been able to get a polyfill working? I've tried this plugin I built, but still no luck: 

```tsx
function alsShim() {
  return {
    name: 'virtual-async-hooks',
    resolveId(id) {
      if (id === '\0virtual:async_hooks') return id;
    },
    load(id) {
      if (id !== '\0virtual:async_hooks') return null;

      console.log('loaded!', id);

      // Browser-friendly shim
      return `
        import AsyncContext from "@webfill/async-context";

        console.log('hello!')

        export class AsyncLocalStorage {
          #var = new AsyncContext.Variable({ defaultValue: undefined });
          
          run(store, fn, ...args) {
            console.log('Run', store)
            return this.#var.run(store, fn, ...args);
          }
          getStore() {
            console.trace('Get Store', this, this.#var.get())
            return this.#var.get();
          }
          enterWith(store) {
            this.#var.run(store, () => {});
          }
          disable() {}
          enable() {}
        }

        export default { AsyncLocalStorage };
      `;
    },
  };
}
```

**Comment by SamVerschueren** [comment]
So I worked on AsyncLocalStorage support for regular promise chains. So for your example, if you use `.then()` instead of `async/await`, this will now work as you can see in this example https://stackblitz.com/~/edit/stackblitz-starters-4j9cg977.

We can't make this work with native `async/await` because that is handled by the browser and we can't know when a promise is scheduled or resolved etc. in order to call the correct hooks for ASL to work.

The only way we can make it work with async/await is if we transpile them, which is also not great because that comes with a cost.

**Comment by birkskyum** [comment]
The async/await are very widely adopted, so any effort towards supporting it would remove a big barrier. It's a bit of a non-starter right now that it's msising.

**Comment by 0xSufi** [comment]
Any progress on this?
